source/tutorials/nix-language.md

```diff
@@ -0,0 +1,1600 @@
+
+```nix
+let
+  f = x: x + 1
+in f
+```
+
+    <LAMBDA>
+
+### Calling functions
+
+Also known as "function application".
+
+Calling a function with an operand means writing the operand after the function.
+
+Example:
+
+```nix
+(x: x + 1) 1
+```
+
+    2
+
+:::{note}
+Since function and operand are separated by white space, sometimes parantheses (`( )`) are required to distinguish expressions.
+:::
+
+Example:
+
+```nix
+let
+  f = x: x.a;
+in
+f { a = 1; }
+```
+
+    1
+
+The above example calls `f` on a literal attribute set.
+One can also pass operands by name.
+
+Example:
+
+```nix
+let
+  f = x: x.a;
+  v = { a = 1; };
+in
+f v
+```
+
+    1
+
+
+### Chaining arguments
+
+Arguments can be chained by nesting functions.
+
+Such a nested function can be used like a function that takes multiple arguments, but offers additional flexibility.
+
+Example:
+
+```nix
+x: y: x + y
+```
+
+    <LAMBDA>
+
+The above function takes one argument and returns a function `y: x + y` with `x` set to the value of the argument.
+
+Example:
+
+```nix
+let
+  f = x: y: x + y;
+in
+f 1
+```
+
+    <LAMBDA>
+
+
+Applying that function `f 1` to another argument yields the inner body `x + y` (with `x` set to `1` and `y` set to the other argument), which can now be fully evaluated.
+
+```nix
+let
+  f = x: y: x + y;
+in
+f 1 2
+```
+
+    3
+
+<!-- TODO: exercise - assign the lambda a name and do something with it -->
+
+### Attribute set argument
+
+Also known as “keyword arguments” or “destructuring” .
+
+Nix functions can be declared to require an attribute set with specific structure as argument.
+
+This is denoted by listing the expected attribute names separated by commas (`,`) and enclosed in braces (`{ }`).
+
+Example:
+
+```nix
+{a, b}: a + b
+```
+
+    <LAMBDA>
+
+The argument defines the exact attributes that have to be in that set.
+Leaving out or passing additional attributes is an error.
+
+Example:
+
+```nix
+let
+  f = {a, b}: a + b;
+in
+f { a = 1; b = 2; }
+```
+
+    3
+
+Counter-example:
+
+```nix
+let
+  f = {a, b}: a + b;
+in
+f { a = 1; b = 2; c = 3; }
+```
+
+    error: 'f' at (string):2:7 called with unexpected argument 'c'
+
+           at «string»:4:1:
+
+                3| in
+                4| f { a = 1; b = 2; c = 3; }
+                 | ^
+                5|
+
+
+<!-- TODO: not the same as x: x.a + x.b (!!!!) -->
+
+### Default attributes
+
+Also known as “default arguments”.
+
+Destructured arguments can have default values for attributes.
+
+This is denoted by separating the attribute name and its default value with a question mark (`?`).
+
+Attributes in the argument are not required if they have a default value.
+
+Example:
+
+```nix
+let
+  f = {a, b ? 0}: a + b
+in
+f { a = 1; }
+```
+
+    1
+
+Example:
+
+```nix
+let
+  f = {a ? 0, b ? 0}: a + b
+in
+f { } # empty attribute set
+```
+
+    0
+
+### Additional attributes
+
+Additional attributes are allowed with an ellipsis (`...`):
+
+    {a, b, ...}: a + b
+
+Example:
```

Comment:
Just read through this and this section was the only thing to give me pause. I already knew this but I think bringing a little more attention to the use of ellipsis would be good.

```suggestion
Unlike the previous counter-example, passing additional arguments will not be an error.

Example:
```